src/Types.hs

```diff
@@ -148,14 +152,87 @@ miThreadTagsEditor = lens _miThreadTagsEditor (\m v -> m { _miThreadTagsEditor =
 miNewMail :: Lens' MailIndex Int
 miNewMail = lens _miNewMail (\m v -> m { _miNewMail = v})
 
+-- | Type representing a specific entity from an e-mail for display.
+--
+newtype MailBody =
+  MailBody [Paragraph]
+  deriving (Show, Eq)
+
+mbParagraph :: Traversal' MailBody Paragraph
+mbParagraph f (MailBody xs) = fmap (\xs' -> MailBody xs') (traverse f xs)
+
+matchCount :: MailBody -> Int
+matchCount =
+  foldrOf
+    (mbParagraph . pLine)
+    (\l amount -> view (lMatches . to length) l + amount)
+    0
+
+-- | A paragraph in the mail body
+--
+newtype Paragraph =
+  Paragraph [Line]
+  deriving (Show, Eq)
+
+pLine :: Traversal' Paragraph Line
+pLine f (Paragraph xs) = fmap (\xs' -> Paragraph xs') (traverse f xs)
+
+-- | A match of a substring in the current line of text
+--
+data Match =
+  Match Int -- ^ offset
+        Int -- ^ length
+        Int -- ^ line number
+  deriving (Show, Eq)
+
+mLinenumber :: Lens' Match Int
+mLinenumber f (Match a b c) = fmap (\c' -> Match a b c') (f c)
+
+-- | A scroll step indicated by the sequential number, line number and
+-- the match.
+-- The sequential number is used for visual purposes to render a
+-- status like: 2 of 30 matches
+--
+type ScrollStep = (Int, Int, Match)
+
+stNumber :: Lens' ScrollStep Int
+stNumber f (a, b, c) = fmap (, b, c) (f a)
```

Comment:
FYI, see `_1` (and for `stMatch` below, `_3`) from http://hackage.haskell.org/package/lens-4.18.1/docs/Control-Lens-Tuple.html.

Reply:
Aah yes of course. Why have I not thought of that. Will change.